### PR TITLE
applications: nrf5340_audio: Implement parts of CSIP

### PR DESCRIPTION
**Pull request by alexsven**
- Use SIRK to discover and connect to other members of the same set.
- If device is not part of set, scan by device name as normal.
- Move restart of scanning after connected to one device to main.
- OCT-3024

### CODE REVIEW

**Review by gWacey — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c [hunk 1]
```diff
@@ -156,6 +162,62 @@ static bool device_name_check(struct bt_data *data, void *user_data)
...
+		struct bt_conn_info info;
+		struct bt_conn *existing_conn = bt_conn_lookup_addr_le(BT_ID_DEFAULT, addr);
+
+		if (existing_conn != NULL) {
+			ret = bt_conn_get_info(existing_conn, &info);
+			if (ret == 0 && info.state == BT_CONN_STATE_CONNECTED) {
+				LOG_DBG("Trying to connect to an already connected conn");
+				bt_conn_unref(existing_conn);
+				return false;
+			}
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:183]: Should we print a WRN if `ret != 0` ?

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c [hunk 2]
```diff
@@ -156,6 +162,62 @@ static bool device_name_check(struct bt_data *data, void *user_data)
...
+			/* Unref is needed due to bt_conn_lookup */
+			bt_conn_unref(existing_conn);
+		}
+		LOG_DBG("Coordinated set device found");
+		sirk_local = NULL;
+
+		bt_le_scan_cb_unregister(&scan_callback);
+		cb_registered = false;
+
+		ret = bt_le_scan_stop();
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:194]: Should we stop the scan before doing `bt_le_scan_cb_unregister` ? What happens if we get a scan result between line 191 and 194?

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c [hunk 3]
```diff
@@ -156,6 +162,62 @@ static bool device_name_check(struct bt_data *data, void *user_data)
...
+		if (ret) {
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:195]: If I am reading correctly, we stop the scan after a set member is found. A set can contain N members, so this looks like the scan may stop too soon.

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c [hunk 4]
```diff
@@ -182,7 +244,11 @@ static void scan_recv_cb(const struct bt_le_scan_recv_info *info, struct net_buf
 	case BT_GAP_ADV_TYPE_SCAN_RSP:
 		/* Note: May lead to connection creation */
 		if (bonded_num < CONFIG_BT_MAX_PAIRED) {
-			bt_data_parse(ad, device_name_check, (void *)info->addr);
+			if (sirk_local == NULL) {
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:247]: What is meant by `sirk_local` on the unicast client side?

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c [hunk 5]
```diff
@@ -195,6 +261,43 @@ static void scan_recv_cb(const struct bt_le_scan_recv_info *info, struct net_buf
 	}
 }
 
+static void conn_in_set_check(struct bt_conn *conn, void *data)
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:264]:
```suggestion
static void conn_in_coord_set_check(struct bt_conn *conn, void *data)
```

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c [hunk 6]
```diff
@@ -195,6 +261,43 @@ static void scan_recv_cb(const struct bt_le_scan_recv_info *info, struct net_buf
...
+{
+	int ret;
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:266]: Should do some NULL checks here.

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c [hunk 7]
```diff
@@ -195,6 +261,43 @@ static void scan_recv_cb(const struct bt_le_scan_recv_info *info, struct net_buf
...
+
+	uint8_t *num_filled = (uint8_t *)data;
+
+	ret = bt_conn_get_info(conn, &info);
+	if (ret) {
+		LOG_ERR("Failed to get conn info for %p: %d", (void *)conn, ret);
+		return;
+	}
+
+	/* Don't care about connection not in a connected state */
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:278]: Why?

File: applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c [hunk 8]
```diff
@@ -1252,13 +1253,34 @@ static void unicast_discovery_complete_cb(struct bt_conn *conn, int err,
...
 					  const struct bt_csip_set_coordinator_csis_inst *csis_inst)
 {
-	if (err) {
+	int ret;
+	struct le_audio_msg msg;
+	uint8_t index;
+
+	ret = device_index_get(conn, &index);
+	if (ret) {
+		return;
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c:1262]: Add a print?

File: applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c [hunk 9]
```diff
@@ -1252,13 +1253,34 @@ static void unicast_discovery_complete_cb(struct bt_conn *conn, int err,
...
+	}
+
+	if (err || csis_inst == NULL) {
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c:1265]: Check for err at the beginning of this function. If there is an err, it may be impossible to do `device_index_get`

File: applications/nrf5340_audio/unicast_client/main.c [hunk 10]
```diff
@@ -292,6 +292,40 @@ static void le_audio_msg_sub_thread(void)
...
+				LOG_DBG("Setting SIRK");
+				bt_mgmt_scan_sirk_set(msg.sirk);
+				bt_mgmt_set_size_filled_get(&num_filled);
+
+				LOG_INF("Set members found so far: %d/%d", num_filled,
+					msg.set_size);
+
+				if (num_filled == msg.set_size) {
+					/* All devices in set found, clear sirk before scanning */
+					bt_mgmt_scan_sirk_set(NULL);
```
Comment (koffes) [inline applications/nrf5340_audio/unicast_client/main.c:312]: Why is this required?

Comment (koffes) [reply applications/nrf5340_audio/unicast_client/main.c:312]: I may have missed it, but if one of the set members disconnect, where is this reset?

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c [hunk 11]
```diff
@@ -195,6 +261,43 @@ static void scan_recv_cb(const struct bt_le_scan_recv_info *info, struct net_buf
...
+	}
+}
+
+void bt_mgmt_set_size_filled_get(uint8_t *num_filled)
+{
+	bt_conn_foreach(BT_CONN_TYPE_LE, conn_in_set_check, (void *)num_filled);
+}
+
+void bt_mgmt_scan_sirk_set(const uint8_t *sirk)
+{
```
[the remark this answers is not in the document]
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:297]: Yeah, I had it one way first then I changed my mind :P Will make it consistent

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c (hunk 2 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:194]: It shouldn't matter. If we get a scan result and there is cb then nothing will happen

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c (hunk 3 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:195]: We always stop scan before we start to create a connection since there is a  lot of data going back and forth during connection. Scanning is restarted after the device is connected if the set is not complete and we haven't hit the limit of CONFIG_BT_MAX_CONN

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c (hunk 4 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:247]: It is just a way to temporary store the sirk of the set we are currently looking for. Since we are using the `bt_conn_foreach()` we need to store the pointer locally. We could use a struct and then pack the num_filled and sirk in there, it just seemed easier this way.

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c (hunk 6 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:266]: We could check that data is not NULL, but conn we do not need to check since this is provided by host through it's conn list.

File: applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c (hunk 7 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/scanning/bt_mgmt_scan_for_conn.c:278]: Because we stop scanning when connection to a device, so we know that the only valid states that we care about is `CONNECTED`. Also if we find a device in `DISCONNECTED` state it shouldn't count as a set member.

File: applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c (hunk 8 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c:1262]: There is already a print in the device_index_get

File: applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c (hunk 9 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c:1265]: No, the err usually indicates that the discovery didn't complete successfully, meaning potentially that the device doesn't have CAS. In can still be connected, especially after this PR is merged and upmerged: https://github.com/zephyrproject-rtos/zephyr/pull/73475

File: applications/nrf5340_audio/unicast_client/main.c (hunk 10 above)
Comment (alexsven) [reply applications/nrf5340_audio/unicast_client/main.c:312]: When the sirk is set we will only accept devices from that set. When setting it to NULL we accept any device (like we did before). 

So the mechanism now is that when we connect to a device that is part of a set, we will only look for its partner. If it's not part of a set we will accept any device with the same device name

File: applications/nrf5340_audio/src/bluetooth/bt_management/bt_mgmt.h [hunk 12]
```diff
@@ -55,6 +55,22 @@ enum bt_mgmt_scan_type {
...
+ *
+ * @param[out]	num_filled	The number of connected set members.
+ */
+void bt_mgmt_set_size_filled_get(uint8_t *num_filled);
+
+/**
+ * @brief	Set 'Set Identity Resolving Key' (SIRK).
+ *		Used for searching for other member of the same set.
+ *
+ * @param[in]	sirk	Pointer to the Set Identity Resolving Key to store.
```
[the remark this answers is not in the document]
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/bt_mgmt.h:70]: Done

File: applications/nrf5340_audio/src/bluetooth/bt_management/bt_mgmt.h [hunk 13]
```diff
@@ -109,6 +125,13 @@ int bt_mgmt_manufacturer_uuid_populate(struct net_buf_simple *uuid_buf, uint16_t
 int bt_mgmt_adv_start(const struct bt_data *ext_adv, size_t ext_adv_size,
 		      const struct bt_data *per_adv, size_t per_adv_size, bool connectable);
 
+/**
+ * @brief	Get the number of active connections.
+ *
+ * @param[out]	num_conn	The number of active connections.
```
[the remark this answers is not in the document]
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/bt_mgmt.h:131]: Seemed a bit much to write:
'Pointer to the area to populate with the number of active connections'

File: applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c (hunk 9 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_stream/unicast/unicast_client.c:1265]: As discussed device_index_get should always return an index here since the device has been added to the unicast_servers array in the function that started the discovery process.

File: applications/nrf5340_audio/src/bluetooth/bt_management/bt_mgmt.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (gWacey) [reply applications/nrf5340_audio/src/bluetooth/bt_management/bt_mgmt.h:131]: Okay, but this forms part of the documentation and we would normally describe it this way.

[the remark this answers is not in the document]
Comment (alexsven) [reply applications/nrf5340_audio/src/bluetooth/bt_management/bt_mgmt.h:131]: Fair enough, will change it to:
```suggestion
 * @param[out]	num_conn	Pointer to populate with the number of active connections found.
```

**Review by koffes — APPROVED** [review]
